test/integration/isolated/udpingress_tcpingress_collision_test.go

```diff
@@ -0,0 +1,197 @@
+//go:build integration_tests
+
+package isolated
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"strconv"
+	"syscall"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/kong/kubernetes-testing-framework/pkg/clusters"
+	ktfkong "github.com/kong/kubernetes-testing-framework/pkg/clusters/addons/kong"
+	"github.com/kong/kubernetes-testing-framework/pkg/utils/kubernetes/generators"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/annotations"
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1beta1"
+	"github.com/kong/kubernetes-ingress-controller/v3/pkg/clientset"
+	"github.com/kong/kubernetes-ingress-controller/v3/pkg/clientset/typed/configuration/v1beta1"
+	"github.com/kong/kubernetes-ingress-controller/v3/test"
+	"github.com/kong/kubernetes-ingress-controller/v3/test/integration/consts"
+	"github.com/kong/kubernetes-ingress-controller/v3/test/internal/testlabels"
+)
+
+func TestUDPIngressTCPIngressCollision(t *testing.T) {
+	// Constants shared in many steps of this test that doesn't change.
+	const (
+		serviceName         = "udp-and-tcp-echo"
+		commonTCPandUDPPort = 7777
+		udpIngressName      = "udp-echo-ingress"
+		tcpIngressName      = "tcp-echo-ingress"
+	)
+	testUUID := uuid.NewString()
+
+	// Helpers used in this test.
+	requireNoResponse := func(t *testing.T, protocol test.Protocol, udpGatewayURL string) {
+		t.Helper()
+		assert.EventuallyWithT(t, func(c *assert.CollectT) {
+			err := test.EchoResponds(protocol, udpGatewayURL, "irrelevant")
+			switch protocol {
+			case test.ProtocolUDP:
+				// For UDP lack of response (a timeout) means that we can't reach a service.
+				assert.True(c, os.IsTimeout(err), "unexpected error for UDP: %v", err)
+			case test.ProtocolTCP:
+				assert.True(c, errors.Is(err, io.EOF) || errors.Is(err, syscall.ECONNRESET), "unexpected error for TCP: %v", err)
+			default:
+				assert.FailNow(c, "unsupported protocol: %v", protocol)
+			}
+		}, consts.IngressWait, consts.WaitTick)
+	}
+	requireResponse := func(t *testing.T, protocol test.Protocol, udpGatewayURL, expectedMsg string) {
+		t.Helper()
+		assert.EventuallyWithT(t, func(c *assert.CollectT) {
+			assert.NoError(c, test.EchoResponds(protocol, udpGatewayURL, expectedMsg))
+		}, consts.IngressWait, consts.WaitTick)
+	}
```

Comment:
I'm seeing those assertion helpers for the nth time I believe. Couldn't we share those between tests?

Reply:
I think so too, but I deliberately waited for the last PR to see all the usecases,I'm going to address this in a separate PR

Reply:
Alright. Good to know that you have this planned. I suggest we start using those and evolve as we go.